### PR TITLE
feat(promise): add anyP

### LINKED ISSUE #807: anyP
Make sense to wait untill stage-2 is reached. https://github.com/tc39/proposal-promise-any

**Update:**
Already present in other Promise libraries. No need to wait for stage-2.

### PR DESCRIPTION
**Pull request by tynopet**
Closes #807.

### CODE REVIEW

File: src/anyP.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import curry1 from 'ramda/src/internal/_curry1';
+
+import isFunction from './isFunction';
+import polyfill, { AggregatedError } from './internal/polyfills/Promise.any';
+
+export const anyPPolyfill = curry1(polyfill);
+export { AggregatedError };
+
+/**
+ * returns a promise that is fulfilled by the first given promise to be fulfilled,
```
Comment (char0n) [inline src/anyP.js]: returns -> Returns

File: src/internal/polyfills/Promise.any.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  return new Promise((resolve, reject) => {
+    const onReject = e => {
+      exceptions.push(e);
+
+      if (exceptions.length === iterable.length) {
+        reject(new AggregatedError(exceptions));
+      }
+    };
+
+    map(
```
Comment (char0n) [inline src/internal/polyfills/Promise.any.js]: Suggestion without epxlicit `{}`. Formatting will be handled by prettier.
```js
map(p => resolveP(p).then(resolve).catch(onReject), [...iterable])

File: types/index.d.ts [hunk 3]
```diff
@@ -784,6 +784,12 @@ declare namespace RamdaAdjunct {
         */
         allSettledP<T>(iterable: Iterable<T>): Promise<Array<SettledPromise<T>>>;
 
+        /**
+         * anyP accepts an iterable of promises and returns a promise that is fulfilled by the first given promise to be fulfilled,
```
Comment (char0n) [inline types/index.d.ts]: Doesn't match the description in the source: https://github.com/char0n/ramda-adjunct/pull/1080/files#diff-4cb076891bd6a121df6d7ea58b2d06b7R11. We need to have this identical.

File: test/anyP.js [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+      const v3 = 3;
+      const actual = [v1, v2, v3];
+      const expected = 1;
+
+      assert.strictEqual(await RA.anyP(actual), expected);
+    });
+  });
+
+  context('given list of values and promises', function() {
+    specify('should return first of fulfilled values', async function() {
```
Comment (char0n) [inline test/anyP.js]: This test doesn't correspond with example here: https://github.com/char0n/ramda-adjunct/pull/1080/files#diff-4cb076891bd6a121df6d7ea58b2d06b7R27

File: test/anyP.js [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+    specify(
+      'should fail with AggregatedError that contains a list of rejected reasons',
+      async function() {
+        const p1 = RA.rejectP(1);
+        const p2 = RA.rejectP(2);
+        const p3 = RA.rejectP(3);
+        const actual = [p1, p2, p3];
+        const expected = [1, 2, 3];
+
+        try {
```
Comment (char0n) [inline test/anyP.js]: Checkout the [assert.throw](https://www.chaijs.com/api/assert/#method_throws). It's more idiomatic to the problem you're trying to solve here.

File: types/index.d.ts [hunk 6]
```diff
@@ -784,6 +784,12 @@ declare namespace RamdaAdjunct {
...
+         * or rejected with an array of rejection reasons if all of the given promises are rejected.
+         */
+        anyP<T>(iterable: Iterable<T>): Promise<T>;
```
Comment (char0n) [inline types/index.d.ts]: We're missing tests for typings. You can learn more about how to do them here: https://github.com/char0n/ramda-adjunct/tree/master/types/test

File: test/anyP.js (hunk 5 above)
Comment (tynopet) [reply test/anyP.js]: I tried this, but I couldn't use it. Assert.throw has an argument that is a function but it calls it without arguments and doesn't work with async functions.

File: src/anyP.js (hunk 1 above)
Comment (tynopet) [reply src/anyP.js]: Oh, I'm sorry, I fixed it

File: src/internal/polyfills/Promise.any.js (hunk 2 above)
Comment (tynopet) [reply src/internal/polyfills/Promise.any.js]: I removed the braces, but prettier saved the formating

File: types/index.d.ts (hunk 3 above)
Comment (tynopet) [reply types/index.d.ts]: Thank you, I fixed it

File: test/anyP.js (hunk 4 above)
Comment (tynopet) [reply test/anyP.js]: Thank you, I fixed it

File: types/index.d.ts (hunk 6 above)
Comment (tynopet) [reply types/index.d.ts]: Thank you, I added tests for typings

File: test/anyP.js [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+          await RA.anyP(actual);
+          throw new Error('resolving should fail');
+        } catch (e) {
+          assert.notStrictEqual(e.message, 'resolving should fail');
+          assert.strictEqual(e instanceof AggregatedError, true);
```
Comment (char0n) [inline test/anyP.js]:
```js
assert.instanceOf(e, AggreatedError)
```

Comment (char0n) [reply test/anyP.js]: This is just a suggestion with more concise way of doing the same

File: types/index.d.ts [hunk 8]
```diff
@@ -784,6 +784,12 @@ declare namespace RamdaAdjunct {
         */
         allSettledP<T>(iterable: Iterable<T>): Promise<Array<SettledPromise<T>>>;
 
+        /**
+         * anyP returns a promise that is fulfilled by the first given promise to be fulfilled,
```
Comment (char0n) [inline types/index.d.ts]: This is still different then the description in `anyP.js`. We have a convention to keep the description exactly the same.

```js
/**
 * Returns a promise that is fulfilled by the first given promise to be fulfilled,
 * or rejected with an array of rejection reasons if all of the given promises are rejected.
 *
....

**Comment by char0n** [comment]
Nicely done. If you're interested I exposed more promise related functions with `Hacktoberfest` label: https://github.com/char0n/ramda-adjunct/issues?q=is%3Aissue+is%3Aopen+label%3AHacktoberfest

**Pull request merged**